#! /bin/sh /usr/share/dpatch/dpatch-run
## 01-varnishncsa-udp.dpatch by  <root@ragweed.knams.wikimedia.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: add server name to log output, format more like wmf squidlog

@DPATCH@

--- varnish3-3.0.0~/bin/varnishncsa/varnishncsa.c	2011-08-09 23:55:17.000000000 +0000
+++ varnish3-3.0.0/bin/varnishncsa/varnishncsa.c	2011-09-14 21:52:17.000000000 +0000
@@ -79,6 +79,12 @@
 #include "varnishapi.h"
 #include "base64.h"
 
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netdb.h> 
+
+static long int sequence_number = 0;
 static volatile sig_atomic_t reopen;
 
 static struct logline {
@@ -111,6 +117,8 @@
 
 static const char *format;
 
+struct hostent* h;
+
 static int
 isprefix(const char *str, const char *prefix, const char *end,
     const char **next)
@@ -533,6 +541,8 @@
 	/* We have a complete data set - log a line */
 
 	fo = priv;
+	
+	sequence_number++;
 
 	for (p = format; *p != '\0'; p++) {
 
@@ -559,13 +569,18 @@
 			else
 				fprintf(fo, "%s", lp->df_h ? lp->df_h : "-");
 			break;
+
 		case 'l':
-			fprintf(fo, "-");
+			fprintf(fo, "%s", h->h_name);
 			break;
 
 		case 'm':
 			fprintf(fo, "%s", lp->df_m);
 			break;
+			
+		case 'n':
+			fprintf(fo, " %ld ", sequence_number);
+			break;
 
 		case 'q':
 			fprintf(fo, "%s", lp->df_q ? lp->df_q : "");
@@ -594,7 +609,7 @@
 
 		case 't':
 			/* %t */
-			strftime(tbuf, sizeof tbuf, "[%d/%b/%Y:%T %z]", &lp->df_t);
+			strftime(tbuf, sizeof tbuf, "%Y-%m-%dT%T", &lp->df_t);  /* 2011-08-11T21:17:01 - no ms from strftim */
 			fprintf(fo, "%s", tbuf);
 			break;
 
@@ -695,12 +710,59 @@
 open_log(const char *ofn, int append)
 {
 	FILE *of;
+	int sockfd, portno, n;
+	struct sockaddr_in serv_addr;
+	struct in_addr iaddr;
+	struct hostent *server;
+	u_char ttl = 10;
+	char host[500];
+	char port[10];
+	char loopch=0;
+
+	sscanf(ofn, "%[^:]:%s", host, port);
+
+	portno = atoi(port);
+	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+
+	if (sockfd < 0) {
+        	fprintf(stderr, "ERROR opening socket\n");
+	}
+
+   	server = gethostbyname(host);
+
+        if (server == NULL) {
+		fprintf(stderr,"ERROR, no such host\n");
+		exit(1);
+        }
+
+	bzero((char *) &serv_addr, sizeof(serv_addr));
+
+	serv_addr.sin_family = AF_INET;
+	serv_addr.sin_port = htons(portno);
+	bcopy((char *)server->h_addr, (char *)&serv_addr.sin_addr.s_addr, 
+		server->h_length);	
+	if ((ntohl(serv_addr.sin_addr.s_addr) >> 28) == 0xe) {
+		iaddr.s_addr = INADDR_ANY; /* use DEFAULT interface */
+		setsockopt(sockfd, IPPROTO_IP, IP_MULTICAST_IF, &iaddr,
+			sizeof(struct in_addr));
+		setsockopt(sockfd, IPPROTO_IP, IP_MULTICAST_TTL, &ttl,
+			sizeof(unsigned char));
+		setsockopt(sockfd, IPPROTO_IP, IP_MULTICAST_LOOP, /* don't send to own interface */
+			(char *)&loopch, sizeof(loopch));
+	}
+
+	if (connect(sockfd,(struct sockaddr *) &serv_addr,sizeof(serv_addr)) < 0) {
+		perror("socket");
+		exit(1);
+	}
+
+	FILE *sf = fdopen(sockfd, "w");
 
-	if ((of = fopen(ofn, append ? "a" : "w")) == NULL) {
-		perror(ofn);
+	if (sf == NULL) {
 		exit(1);
 	}
-	return (of);
+
+        return (sf);
 }
 
 /*--------------------------------------------------------------------*/
@@ -711,7 +773,7 @@
 
 	fprintf(stderr,
 	    "usage: varnishncsa %s [-aDV] [-n varnish_name] "
-	    "[-P file] [-w file]\n", VSL_USAGE);
+	    "[-P file] [-w host:port]\n", VSL_USAGE);
 	exit(1);
 }
 
@@ -723,8 +785,9 @@
 	const char *P_arg = NULL;
 	const char *w_arg = NULL;
 	struct vpf_fh *pfh = NULL;
+	char hostname[1024];
 	FILE *of;
-	format = "%h %l %u %t \"%r\" %s %b \"%{Referer}i\" \"%{User-agent}i\"";
+	format = "%l %n %t %{Varnish:time_firstbyte}x %h %{Varnish:handling}x/%s %b %m http://%{Host}i%U%q - - %{Referer}i %{X-Forwarded-For}i %{User-agent}i";
 
 	vd = VSM_New();
 	VSL_Setup(vd);
@@ -739,7 +802,7 @@
 				fprintf(stderr, "-f and -F can not be combined\n");
 				exit(1);
 			}
-			format = "%{X-Forwarded-For}i %l %u %t \"%r\" %s %b \"%{Referer}i\" \"%{User-agent}i\"";
+			format = "%l %n %t %{Varnish:time_firstbyte}x %{X-Forwarded-For} %{Varnish:handling}x/%s %b %m http://%{Host}i%U%q - - %{Referer}i %{X-Forwarded-For}i %{User-agent}i";
 			format_flag = 1;
 			break;
 		case 'F':
@@ -786,6 +849,10 @@
 		}
 	}
 
+	hostname[1023] = '\0';
+	gethostname(hostname, 1023);
+	h = gethostbyname(hostname);
+
 	VSL_Arg(vd, 'c', optarg);
 
 	if (VSL_Open(vd, 1))
